lib/internal/modules/esm/public_loader_proxy.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+const { InternalWorker } = require('internal/worker');
+const path = require('path');
+const { deserialize, serialize } = require('v8');
+const { debug } = require('internal/modules/esm/debug');
+
+
+/**
+ * The communication vehicle constituting all memory shared between the main and the worker
+ */
+const commsChannel = new SharedArrayBuffer(2048);
+/**
+ * The lock/unlock segment of the shared memory. Atomics require this to be a Int32Array. This
+ * segment is used to tell the main to sleep when the worker is processing, and vice versa (for the
+ * worker to sleep whilst the main thread is processing).
+ * 0 → main sleeps
+ * 1 → worker sleeps
+ */
+const lock = new Int32Array(commsChannel, 0, 4);
+/**
+ * The request & response segment of the shared memory. TextEncoder/Decoder (needed to convert
+ * requests & responses into a format supported by the comms channel) reads and writes with
+ * Uint8Array.
+ */
+const requestResponseData = new Uint8Array(commsChannel, 4, 2044);
+
+debug('creating worker for publicESMLoader');
+const worker = new InternalWorker(path.resolve('lib/internal/modules/esm/', 'worker.js'), {
```

Comment:
@aduh95 using `__dirname` in this (CJS) file triggers a ReferenceError. Any idea why?

Reply:
Internal modules are using a different wrapping function than user-land ones so it has access to `primordials`, `internalBindings`: https://github.com/nodejs/node/blob/78508028e3439f6210b4a8cc90ea423ce044e7ac/src/node_builtins.cc#L376-L387

While userland modules have access to `__filename` and `__dirname`: https://github.com/nodejs/node/blob/bf47abe5f366f8cdd036257147a5274b3ed01979/lib/internal/modules/cjs/loader.js#L1088-L1095

Reply:
Ahh, okay! Is one of those isolates a viable alternative to `__dirname`? I think `internalBinding` is just for accessing C stuff. I would rather avoid specifying a path from the project root, especially since the file I want is adjacent.

Reply:
What are you trying to do? You should never need `__dirname` in internal code, is it just for debugging, or?

Reply:
Would `node:internal/modules/esm/worker` work? Obviously using an absolute path won't work once `node` is compiled and the source file is nowhere in the FS.

Reply:
I'm trying to supply the main worker file (aka `filename`). The file is adjacent to the instantiating the worker, but `'./…'` does not work—it results in a filepath relative to the current working directory (project root) instead of the current module. PS I really wish node would change this behaviour—it's _never_ what anyone wants.

Reply:
> Would node:internal/modules/esm/worker work

No, that results in `file:///…/nodejs/node/node:internal/modules/esm/worker.js`; `node:` is literally printed into the file path, and no permutation I could think of results in anything valid.

Reply:
What I'm trying to say is that when `node` binary is compiled, the source file won't be available in the file system, so using `path.resolve`, or `__dirname`, or any kind of path would fail. I think we would need to modify the `WorkerImpl` to make it able to load internal modules.

Reply:
OH 😳 well that might explain why the Worker is not loading the script its passed.

Reply:
I looked into it, we probably need to define a new value for `doEval`, is it OK if I send a commit to your branch?

Reply:
Sure, no problem. Thanks!

Reply:
Thanks for the commit! I pulled it down, and it appears to have a same result (same debugs, and it appears to get stuck at the same point—the worker's script doesn't run). Aside from a cleaner implementation of what it was doing before and some polish, what does it do / is it supposed to do?

Reply:
I didn't get to the point I wanted, but hopefully it's a step in the right direction. It looks like the worker thread doesn't run because it's waiting for some signal from the main thread or something, and the main thread is blocked by `Atomics.wait`. If I comment out line 45 of `lib/internal/modules/esm/public_loader_proxy.js`, the worker code does get executed, so we would need to figure out what async call is pending before blocking the main thread.